[Frontend] -e and -M switches

beccadax opened the pull request:
Adds two switches to the frontend. `-e <line>` adds `<line>` to a virtual file of code which it can run. `-M<module_name>` adds `<module_name>` to the implicit imported modules list. Together, they ~~fight crime~~ let you write Swift one-liners.

Resolves [SR-5860](https://bugs.swift.org/browse/SR-5860). Unlike #18303, this implementation keeps the code in memory, rather than writing it to a temporary file.

beccadax committed:
Make InputFileList a real type

Needed for some minor gymnastics later.

beccadax committed:
[Driver][Frontend] Add a -e switch

-e takes a line of code and puts it in a virtual InputFile called “-e”, then processes it as though it were a physical file on disk. You can use multiple “-e” switches to add several lines. Useful mainly for quick one-line tests—both for users who would otherwise use a REPL or playground, and for compiler developers who would otherwise use a test file or pipe—but also for any operation where you just need to run one or two lines.

beccadax committed:
Add tests for -e switch

beccadax committed:
Add a -M switch to import modules

jrose-apple commented:
Hm. I'm nervous about `-M`, for two reasons:

- We've had the `-import-module` frontend option for a while, mostly to support playgrounds. But I'm hesitant to expose it as a general, supported command line option, because it means you can have source code that doesn't include all the information needed to compile it. That's already true about search paths and about mixed-source targets, but I kind of feel like we shouldn't be adding to that. Maybe we can restrict `-import-module` to either `-e` mode or interpreter mode in general? (`swift` and not `swiftc`)

- I've very much tried to have command line option names for swiftc be fairly self-explanatory while not following patterns for the sake of patterns (`-f<option>` vs. `-m<option>`; yes, I know there's a reason). Single-letter options sort of fly in the face of that, and also can't be un-claimed in the future. For `-e` it seems obvious that the single letter is justified, and there's precedent in other tools (perl, ruby, osascript, sed). But I'm not sure `-M Foundation` or `-MFoundation` is *sufficiently* better than `-import-module Foundation` or even `-e "import Foundation"` to be worth claiming the option.

I'll take a look at the actual code soon.

brentdax commented:
From my experience with Perl (which uses `-M` and `-e` in exactly this way), I think that `-M` will be pretty valuable. A handful of characters really does make a noticeable difference when you're writing the sorts of things you'd write with `-e`.

On the other hand, I do understand your concerns about people using `-M` on ordinary compiles. I think the easiest solution is to have `-M<module_name>` literally be a shorthand for `-e 'import <module_name>`—that would naturally make it essentially useless for other compiles.

beccadax committed:
Make -M only work for -e switches

@jrose_apple is concerned that people might discover -M and use it inappropriately for global imports on ordinary projects. This alters -M so that it generates import statements within the -e code and returns -import-module to being an obscure frontend-only option (albeit one with a clearer help string).

beccadax committed:
Test -M with SwiftShims, not Foundation

Foundation is not automatically available on Linux, so we need a different module that’s guaranteed to be available. I think SwiftShims will be.

brentdax commented:
@swift-ci please smoke test

jrose-apple commented on include/swift/Frontend/FrontendInputsAndOutputs.h:
```diff
@@ -46,6 +46,10 @@ class FrontendInputsAndOutputs {
 
+  /// Contents of -e switches (empty if none). (Used for lifetime
+  /// management; the buffer is accessed through an InputFile.)
+  std::shared_ptr<llvm::MemoryBuffer> LinesToExecute;
```
Nitpick: This should go before `AllInputs` so that it's destroyed after the input files.

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -185,6 +188,9 @@ def I : JoinedOrSeparate<["-"], "I">, Flags<[FrontendOption, ArgumentIsPath]>,
 def I_EQ : Joined<["-"], "I=">, Flags<[FrontendOption, ArgumentIsPath]>,
   Alias<I>;
 
+def M : Joined<["-"], "M">, Flags<[FrontendOption]>,
```
This should be JoinedOrSeparate. I don't think we have any Joined-only arguments except those that end in `=`.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -172,8 +172,10 @@ static void addCommonFrontendArgs(const ToolChain &TC, const OutputInfo &OI,
+  inputArgs.AddAllArgs(arguments, options::OPT_e, options::OPT_M);
   inputArgs.AddAllArgs(arguments, options::OPT_I);
   inputArgs.AddAllArgs(arguments, options::OPT_F, options::OPT_Fsystem);
+  inputArgs.AddAllArgs(arguments, options::OPT_import_module);
```
This one's no longer necessary now that this is not being promoted to a driver option.

brentdax replied:
You know, I actually checked that because I thought I had added it, but the Xcode blame display said it was from an old commit. I looked again just now and it still said that, but I hit Cmd-S and the blame shifted. I wonder what happened.

jrose-apple commented on lib/Frontend/ArgsToFrontendInputsConverter.cpp:
```diff
@@ -97,6 +97,25 @@ bool ArgsToFrontendInputsConverter::readInputFilesFromCommandLine() {
+        lines += module;
+        lines += "\n";
+      }
+      lines += "#sourceLocation(file: \"-e\", line: 1)\n";
```
Clever!

jrose-apple commented on lib/Frontend/ArgsToFrontendInputsConverter.cpp:
```diff
@@ -158,6 +177,15 @@ ArgsToFrontendInputsConverter::createInputFilesConsumingPrimaries(
+      primaryFiles.erase("-e");
+
+    Files.remove("-e");
+    result.LinesToExecute = LinesToExecute;
```
I'm confused as to what this is doing, but I think it's okay to treat `-e` mode as whole-module, i.e. "no primary files".

davidungar replied:
This block of code suggests to me that there is probably a better approach. What do -e arguments have to do with primary files? Something is not quite right to my eyes here. I *think* this code is using a filename of "-e" as a placeholder for code kept in `LinesToExecute`. But I'm thinking a far better approach would be to generalize `InputFile` to `InputFileOrLineToExecute`, since that's the model you're going after. And put the code to execute right in the `InputFileOrLineToExecute` structure, along with the corresponding modules needed. How does this sound?

Keeping redundant information in both `LinesToExecute` and the InputFile names (-e) feels like it's asking for trouble downstream, sorry.

brentdax replied:
They probably don't have much to do with primary files—I'm probably just confused about when there is a primary file and when there isn't.

`FrontendInputsAndOutputs:: LinesToExecute` is entirely for lifetime management—it is never accessed for anything else. If I comment it out and change the `LinesToExecute` in `ArgsToFrontendInputsConverter` to use a `unique_ptr`, the only difference is that I have to leak the buffer instead of releasing it cleanly.

We ultimately want the lines to execute to be turned into a single `llvm::MemoryBuffer` managed by `SourceManager`. I found `ArgsToFrontendInputsConverter` to be a natural place to concatenate the strings and form a buffer, and `InputFile::Buffer` to be a natural way to get that buffer into the compiler, but we could of course do it in some other way.

davidungar replied:
It's *either* `LinesToExecute` that bother me. The problem is that `LinesToExecute` may have no content, yet there may be an InputFile with a "-e" name. Or there may be content in `LinesToExecute` and no `InputFile` with a "-e". Or many. Yes, one can add assertions, and checks, and diagnostics, but it seems there must be a better way.
For example, if any given frontend job should have either input files or only one -e, then there could be a class triad with a common superclass, one subclass to hold the input files, and a sibling subclass holding the lines to execute. Let's find a way to make code more directly represent what we want it to do.

brentdax replied:
I realized today that the underlying problem here is that if `InputFile` has a `Buffer`, it doesn't take ownership of it. So I think the most general solution is actually to make `InputFile::Buffer` be a `std::shared_ptr` instead of a plain pointer. Is that an acceptable solution? The vast majority of `InputFile`s do not include a `Buffer`; the exceptions are:

* Code generated by the Extract Function refactoring.
* Code generated by applying fix-its in the migrator.
* The main input of `sil-opt`, `sil-func-extractor`, `sil-llvm-gen`, and `sil-nm`.
* `-e` processing.

In all other cases, `Buffer` would be `nullptr`, which I believe would save us from a lot of the overhead that would usually keep us from using a `shared_ptr`.

Does that make sense to you? Or should we try to find some way to transfer ownership of the `Buffer` to a long-lived object more quickly?

jrose-apple commented on test/Driver/lines_to_execute.swift:
```diff
@@ -0,0 +1,26 @@
+// CHECK-4: -e:2:1: error: use of unresolved identifier 'foo'
+// CHECK-4-NOT: Hello, world!
+
+// RUN: %swift_driver_plain -MSwiftShims -e '_stdlib_write(1, "Hello, world!", 13)' 2>&1 | %FileCheck -check-prefix CHECK-5 %s
```
Please don't use SwiftShims in tests. You can make some dummy C module instead.

jrose-apple commented on test/Driver/lines_to_execute.swift:
```diff
@@ -0,0 +1,26 @@
+// RUN: %swift_driver_plain -e 'print("Hello, world!")' 2>&1 | %FileCheck -check-prefix CHECK-1 %s
```
These should just use `%swift_driver` rather than `%swift_driver_plain` so that they still get the right target and module cache. You'll also need `REQUIRES: swift_interpreter` so that they don't fail on non-host tests or platforms where the JIT isn't set up.

jrose-apple commented on test/Driver/lines_to_execute.swift:
```diff
@@ -0,0 +1,26 @@
+// CHECK-7: -e:1:1: error: use of unresolved identifier 'fnord'
+
+// RUN: not %swift_driver_plain -MSwiftShims 2>&1 | %FileCheck -check-prefix CHECK-8 %s
+// CHECK-8: error: cannot use -M arguments to import modules without -e arguments to use them
```
Can you check that `-M` and `-e` are rejected with `%swiftc_driver`, and that you can't mix `-e` and regular files?

brentdax replied:
`-e` happens to *usually* not work with other files (it causes "expressions are not allowed at the top level" errors), but I'll add a specific, reliable diagnostic for that case.

I'd like `-e` to work with at least the frontend modes you might use when debugging how the compiler handles small code samples (like `-dump-ast`, `-emit-sil`, etc.) Just glancing at the `ActionType` enum, here's what I think we should support:

* Parse
* ResolveImports
* Typecheck
* DumpParse
* EmitSyntax
* DumpAST
* PrintAST
* DumpScopeMaps
* DumpTypeRefinementContexts
* EmitImportedModules
* EmitPCH
* EmitSILGen
* EmitSIL
* EmitSIBGen
* EmitSIB
* Immediate
* EmitAssembly
* EmitIR
* EmitBC
* DumpTypeInfo

And here's what we should prevent:

* NoneAction
* DumpInterfaceHash
* EmitModuleOnly
* MergeModules
* REPL (it might actually be a good idea to allow use with the REPL to set up an environment, but that'll require additional work, and we'd want to support files for this purpose at the same time)
* EmitObject

But this is just off the top of my head, and I don't know much about a few of these modes (like DumpInterfaceHash). Is there anything you'd change about these lists, or do you think this is totally the wrong approach?

jrose-apple reviewed (commented):
Okay, I guess I can live with `-M` if it's limited to `-e` mode. Here's the code comments I promised!

davidungar reviewed (changes requested):
Learning this code base can be like the death of a thousand cuts. Each less-than-as-clear-as-we-can-make-it is minor, but together they can add up to much more effort. So, I'm requesting some changes to make things as clear as possible. Happy to discuss them!

After I wrote the above, I realized that the redundancy between `LinesToExecute` and the "-e" file names worries me quite a bit. Can you come up with a design to centralize this information? Thanks!

davidungar commented on lib/Frontend/FrontendInputsAndOutputs.cpp:
```diff
@@ -45,6 +46,7 @@ operator=(const FrontendInputsAndOutputs &other) {
   for (InputFile input : other.AllInputs)
     addInput(input);
   IsSingleThreadedWMO = other.IsSingleThreadedWMO;
+  LinesToExecute = other.LinesToExecute;
```
My bad! The duplication between line 40 and 49 makes my failure to factor `FrontendInputsAndOutputs::FrontendInputsAndOutputs(const FrontendInputsAndOutputs &other)` with `FrontendInputsAndOutputs &FrontendInputsAndOutputs::operator=(const FrontendInputsAndOutputs &other)` stick out like a sore thumb. @jrose-apple do you have a good suggestion for how to factor those duplicated 4 lines? I guess a private static function could work.

jrose-apple replied:
Since we're not too concerned about performance, it's usually possible to implement the copy-constructor as

```cxx
FrontendInputsAndOutputs::FrontendInputsAndOutputs(const FrontendInputsAndOutputs &other) {
  *this = other;
}
```

But if we did care about the cost of default initialization, then there's really no way to do this. C++ just doesn't provide it.

davidungar commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -429,7 +429,7 @@ StringRef SupplementaryOutputPathsComputer::
     return outputFilename;
 
-  if (input.isPrimary() && input.file() != "-")
+  if (input.isPrimary() && input.file() != "-" && input.file() != "-e")
```
Both lines 186 and 432 test if the file is "-" or is "-e". This should be factored. For that matter, I'm not comfortable with the use of '-e' for a filename, since it isn't a filename. (I know that it seems consistent with the use of '-', sigh.) At the very least, there needs to be a Doxygen comment where the `file` member function is defined and where the `InputFile.FileName` instance variable is defined. Better still would be a Kind enumeration for `InputFile`, sigh.

jrose-apple replied:
We need to show *something* in diagnostics, but it wouldn't have to be "-e". We could do something like "<command-line>" instead, as long as it's very unlikely to be an actual filename. 

(And whatever it is, if it's referenced in more than one place I suppose it should probably be a string constant declared somewhere rather than a magic string. "-" is short enough and part of a cross-program idiom that we can probably get away with it, but "-e" is more specific.)

brentdax replied:
Actually, since we're using `#sourceLocation` in some cases anyway, the name we use for that won't necessarily show up in diagnostics. We can still have diagnostics say "-e" without actually using that for the filename.

davidungar commented on lib/Frontend/ArgsToFrontendInputsConverter.h:
```diff
@@ -45,6 +45,9 @@ class ArgsToFrontendInputsConverter {
   const llvm::opt::ArgList &Args;
 
+  /// The code provided in -e switches.
+  std::shared_ptr<llvm::MemoryBuffer> LinesToExecute;
```
See my comment for `createInputFilesConsumingPrimaries`.

brentdax commented on include/swift/Driver/Util.h:
```diff
@@ -29,7 +29,13 @@ namespace driver {
+    bool empty() const {
+      return !hasLinesToExecute && typesAndArgs.empty();
+    }
+  };
```
Pardon my likely ignorance, but what makes this better than having `InputFileList` parameters be `const` in functions where they should not be changed? I'm sure there's a reason; I'm just not seeing it yet.

davidungar replied:
Thanks for asking! If you make `typesAndArgs` and `hasLinesToExecute` `const`, then a reader only has to read those two (adjacent) declarations to see that those values never change after the instance is created. But if you make `InputFileList` parameters `const` in functions where they should not be changed, then the reader *has to find every* place where the instance is passed. Does that help?

CodaFi commented:
@brentdax Just checking on this one since it seems to have gone stale: do you still want to pursue this, or should we release the SR to an external contributor and point them at this PR?

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
